core/application.go

```diff
@@ -93,12 +103,62 @@ func (app *Application) CreateUser(clientID string, id string, externalID *strin
 	return user, nil
 }
 
+func (app *Application) setupSyncManagedGroupTimer() {
+	log.Println("setupSyncManagedGroupTimer")
+
+	//cancel if active
+	if app.syncManagedGroupsTimer != nil {
+		app.syncManagedGroupsTimerDone <- true
+		app.syncManagedGroupsTimer.Stop()
+	}
+
+	app.syncManagedGroups()
+}
+
+func (app *Application) syncManagedGroups() {
+	log.Println("syncManagedGroups")
+	if app.config == nil {
+		return
+	}
+
+	for _, clientID := range app.config.SupportedClientIDs {
+		configs, err := app.storage.FindManagedGroupConfigs(clientID)
+		if err != nil {
+			log.Printf("error finding managed group configs for clientID %s\n", clientID)
+		}
+		if len(configs) > 0 {
+			err = app.synchronizeAuthman(clientID, configs)
+			if err != nil {
+				log.Printf("error syncing authman groups %s\n", err.Error())
+			}
+		}
+	}
+
+	durationMins := 1440
+	if app.config.SyncManagedGroupsPeriod != 0 {
+		durationMins = app.config.SyncManagedGroupsPeriod
+	}
+	duration := time.Minute * time.Duration(durationMins)
+	app.syncManagedGroupsTimer = time.NewTimer(duration)
+	select {
+	case <-app.syncManagedGroupsTimer.C:
+		// timer expired
+		app.syncManagedGroupsTimer = nil
+
+		app.syncManagedGroups()
+	case <-app.syncManagedGroupsTimerDone:
+		// timer aborted
+		app.syncManagedGroupsTimer = nil
+	}
+}
+
```

Comment:
That is a very good point, thank you for catching this! I agree that we do need to prevent multiple instances from attempting to sync at the same time, however I do not agree that this cannot be accomplished using the database. We have used a similar mechanism in the past with good results.

My suggestion is to create a `configs` collection that stores timestamps for `managed_group_sync_started` and `managed_group_sync_ended`. Whenever the timer goes off, this document will be read from the db and the start date will be compared to the current time to ensure that it was more than the configured update period ago. If it was, then the document will be updated with a new start date. Note that these two operations will be performed as a transaction so that they are atomic, meaning that any other instances looking to read at the same time will have to wait until the previous instance has updated this config. Once the sync operation is complete, the document will be updated with the new end time.

Note that we actually had a similar issue anyway for preventing multiple manually triggered syncs through the admin API. There is a flag in memory that is used to return an error if simultaneous syncs are attempted. This will not prevent 2 instances from attempting to sync simultaneously. With the new configs, we can check these timestamps again to ensure that there is not a sync currently in progress on another instance.

I will take a pass at this, but please let me know if you see any issues or have additional concerns about this approach.

Thanks again!

Reply:
I'm fine with such an approach unless we need to enter into workarounds spiral. The main goal is to prevent conflicts and write locks which will put plenty of confused logs.

Reply:
Actually this may be the better way, because I haven't seen yet a working variant with AWS EventBridge on production :)

Reply:
Yep, thanks! The goal is to avoid depending on external resources for this functionality. Hopefully it will make things clearer and easier to maintain